Implement automatic HTML generation

bruxisma opened the pull request:
Add automation via GitHub Actions to generate the specification HTML and
deploy the same to GitHub Pages once merged.

Additionally, rework HTML generation to use Poetry[^1] to set up and
manage a virtual environment used to generate the HTML. This helps
ensure that both local and automated builds are using a uniform
environment.

Note that deployment eschews the `upload-pages-artifact`[^2] action
because the mechanism used here allows us to bundle the files and
correct their permissions in one command. (Additionally, this is done
via the Makefile, which allows the archive to be created locally as well
as via an Action.)

For clarity, the `_ext` directory is renamed to `_extensions`, and the
extensions therein are now "installed" to the Poetry environment.
Additionally, the minimum Sphinx version is now to 6.2 or later, as
that's what's been in use recently and Poetry allows us to be less
"stuck" on what's provided by distributions.

Lastly, the Makefile is now cross platform and it will now work on
Windows as long as a GNU Make (or GNU Make compatible tool) is used.

[^1]: https://python-poetry.org/
[^2]: https://github.com/actions/upload-pages-artifact/

Fixes #40.

bruxisma commented:
One thing to note: once merged, you will need to change the github pages "source" to Github Actions and then kick off a build manually (or change it pre-merge and then it will kick off)

mwoehlke commented:
I'm not that familiar with poetry; will this interfere with using the Makefile to generate the HTML locally? (I'm wondering if it makes sense to change the Makefile and instead use `make` to build, so we aren't duplicating logic.)

I think `sys.path.insert(0, os.path.abspath('_ext'))` needs to be either changed or removed? (BTW, is it possible to keep the name as `_ext`? I think that helps make it clear that it is content related to "rendering" the specification, rather than to the specification itself. Note also `_static`. Yes, I ought to also change the Makefile to `_site`.)

bruxisma commented:
> will this interfere with using the Makefile to generate the HTML locally?

It will not, though I'd recommend using `poetry` via the `Makefile` regardless. What poetry does is provide a lockfile for well known dependencies, the ability to create a virtualenv on a per-project basis (so your global system is left alone), the ability to execute from within said virtualenv without having to run any special scripts, creating an "editable" install that can still be debugged, and making sure anyone who uses it is on the same page. (It also means in your case you're not tied to "whatever fedora provides" but rather "what version of python does fedora provide?"). In other words the documentation build becomes more reproducible and a bit more hermetic.

Also, `_static` et. all. gets rendered verbatim into the build directory. I'll clean this up a bit and move stuff into `_extensions`, but wanted to mention that if we *did* care about how this is all laid out we could easily organize it a bit better. 🙂 

I'll change the `Makefile` to use `_site` as well.

bretbrownjr reviewed (approved):
Great!

I asked for some dev docs on spellcheck dictionary maintenance, but it also makes sense to just make an issue to create some dev docs and circle back on that. Just from this PR, we could use breadcrumbs to:

* sphinx
* poetry
* however spellchecking works

mwoehlke commented:
> I'd recommend possibly putting some approvals for the actual final deployment to GitHub Pages in place

Can you clarify / expand on this?

> Lastly, I took an extra step and added a spellchecker for the github actions steps.

This... runs spell check on the `.rst`s? Is it able to flag only newly-introduced errors? Otherwise I can't imagine, in such a technical work, this producing so many false positives as to make the signal-to-noise ratio unmanageable.

mwoehlke commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+authors = ["The CPS Project <@>"]
+readme = "README.md"
+packages = [
+  { include = "cps.py", from = "_extensions" },
+  { include = "autosectionlabel.py", from = "_extensions" }
```
I've half a mind to leave this as `_ext`, though I suspect "brevity" is the main motivation for that, and thus I feel I shouldn't override you if you feel strongly. WDYT?

bruxisma replied:
I think we should make it clear they're extensions and not maybe any other word that could be confused with it. `_extended` for example.

mwoehlke commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+]
+
+[tool.poetry.dependencies]
+python = "^3.10"
```
BTW, if I only have Python 3.12 locally and try to run `poetry`, does it just use 3.12? Or does it go obtain a local copy of 3.10? (For that matter, what does the `^` mean?)

bruxisma replied:
For information on the version specifications, see poetry's documentation on [dependency specifications](https://python-poetry.org/docs/dependency-specification/)

To succinctly answer your question, `^` means "any newer minor version, but not a newer major version". So as long as Python 4 isn't a requirement we'll be fine 😄 

And yes it'll use your local python. If you want to use different versions of python via poetry, you can install multiple versions of poetry with `pipx` using a different version for each:

```console
$ pipx install poetry --python 3.10 --suffix @3.10
$ poetry@3.10 install --with=docs
$ poetry@3.10 run ...
```

mwoehlke commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+
+[tool.poetry.group.docs.dependencies]
+python = "^3.10"
+sphinx = ">=7.2.6"
```
This suggests we should also update the version requirement in `conf.py`. (Also... 7.2.6? I only have 6.2.1! And I feel like we should maybe pin this to `<8`, unless `>=` does that already?)

mwoehlke replied:
Whee, seems they've been busy making releases! 6.2.1 is less than a year old.

Okay, so maybe we should use `>= 6.2, < 7.3`? Or do we want to allow all the way back to 5.3, as per `conf.py`? If we're effectively raising the minimum here, we should probably change `conf.py` to match.

(Speaking of `conf.py`, I think that `sys.path` tweak still needs updating? :slightly_smiling_face:)

bruxisma replied:
Most of the changes between then and now don't affect us, but we can do that range sure 🙂

bruxisma commented on pyproject.toml:
```diff
@@ -0,0 +1,28 @@
+name = "cps"
+version = "0.dev"
+description = ""
+authors = ["The CPS Project <@>"]
```
The `authors` field is a requirement or poetry won't run. We can always change the name. I honestly just put it there as a placeholder. (Hence the `<@>`, as poetry just checks that there is an `@` to make sure it's a "valid" email address)

mwoehlke replied:
In that case, I suppose we should at least use the mailing list address instead of (null)@(null). :wink:

bruxisma replied:
For what it's worth, this field is only valuable if you're publishing to pypi. Otherwise it *does* sort of run the risk of getting scraped and added to a spam list 😮‍💨

mwoehlke replied:
:shrug: I'm okay with it, then. @bretbrownjr?

mwoehlke replied:
Ah, upon rereading Brett's original comment, IIUC he's also okay with this.

bruxisma commented on .github/workflows/spellcheck.yml:
```diff
@@ -0,0 +1,15 @@
+    branches: [main]
+jobs:
+  spellcheck:
+    name: Spellcheck
```
You would need to open a PR upstream. The dictionary assets are [*here*](https://github.com/crate-ci/typos/tree/master/crates/typos-dict/assets). These get generated into a rust source file and then compiled into the resulting binary.

bruxisma commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+          --verbose
+          --mode=a+rw
+          --file=cps-docs.tar.gz
+          --directory=${{_site
```
oh woops! good catch

bruxisma commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+      id-token: write
+    environment:
+      name: github-pages
+      url: ${{ steps.deployment.outputs.page_url}}
```
yes. it was a copy paste. 😆

mwoehlke replied:
:slightly_smiling_face: I noticed some other workflows seem to like spaces on both sides, so I assume it doesn't matter. Internal consistency is always nice, though.

mwoehlke commented on conf.py:
```diff
@@ -20,7 +20,7 @@
 
 release = '.'.join(map(str, [0, 9, 0]))
-version = re.match('\\d+\.\\d+', release).group(0)
+version = re.match(r'\d+\.\d+', release).group(0)
```
I'd prefer to at least keep the "source of truth" as an int-tuple. There's probably a reason they weren't being set to the same thing? Possibly it relates to the ability to switch between different versions of the doc, which we aren't currently using?

bruxisma replied:
No, it's because of PEP440. Sometimes you have `v1.2.0` versus `v1.2.0b1`

mwoehlke replied:
First, where is it documented that Sphinx can only document projects whose versioning conforms to PEP440?

Second, where is it stated that projects must use the entirety of PEP440?

CPS's version numbers (so far, at least) a) follow semver semantics, and b) look like 'x.y.z', where (x, y, z) is an int-tuple. It happens that this conforms to PEP440, which I'm sure has benefits, but it's incorrect to say that it *is* PEP440. (At best, it's a subset.)

Besides, even if we were to release e.g. `0.11.0b1`, we could write that `'.'.join(map(str, [0, 9, 0])) + 'b1'`. Or we could revisit the decision to keep the canonical representation as an int-tuple at that time.

bruxisma commented:
> > I'd recommend possibly putting some approvals for the actual final deployment to GitHub Pages in place
> 
> Can you clarify / expand on this?

Sure, the documentation is [here](https://docs.github.com/en/actions/managing-workflow-runs/reviewing-deployments). It means that someone can manually approve changes to deploy to github pages.

> > Lastly, I took an extra step and added a spellchecker for the github actions steps.
> 
> This... runs spell check on the `.rst`s? Is it able to flag only newly-introduced errors? Otherwise I can't imagine, in such a technical work, this producing so many false positives as to make the signal-to-noise ratio unmanageable.

I already ran it and there are no detected errors 🙂. This would prevent changes from getting added that have common typos. If a word isn't recognized it rarely results in a false positive (all the same I've had it ignore the css and svg files)

mwoehlke commented on conf.py:
```diff
@@ -4,10 +4,8 @@
 import re
 import sys
 
-sys.path.insert(0, os.path.abspath('_ext'))
```
I think this either needs to be kept (but changed to `_extensions`), or the `Makefile` changed to also use `poetry`? (I think we'd sort-of agreed on the latter?)

bruxisma replied:
I'll fix up the makefile to use poetry over sphinx-build. This way we can at least guarantee everyone is using the same environment and the build is slightly more hermetic 🙂

mwoehlke commented on conf.py:
```diff
@@ -16,11 +14,10 @@
-project = u'Common Package Specification'
-copyright = u'2023, Matthew Woehlke'
+project = 'Common Package Specification'
+copyright = '2023, Matthew Woehlke'
```
If we're touching this line anyway, might as well fix it too; it should be 2024 now...

bruxisma replied:
You can remove technically remove the year. It just keeps copyright in perpetuity until handed over/passed on.

bruxisma replied:
Updated the year and removed the `u` in my merge commit.

mwoehlke commented on conf.py:
```diff
@@ -16,11 +14,10 @@
 
-release = '.'.join(map(str, [0, 9, 0]))
-version = re.match('\\d+\.\\d+', release).group(0)
+release = version = "0.9.0"
```
```suggestion
release = version = '.'.join(map(str, [0, 9, 0]))
```

(Also, we're using `'` for strings everywhere else... but I'd like to keep the int-tuple, please.)

bruxisma replied:
I used `"` because if you'd ever like to use a formatter, `black` is just the way to go.

bruxisma replied:
Sorry for the delayed response to this. I've used what's currently in mainstream, however flake8's developers have become increasingly hostile to the python developer community over the years, refusing to even support the pyproject.toml file and claiming that `pip` needs to change it's entire behavior before they'll even consider supporting it. `ruff`, `black`, and others are more preferable, and if anyone from the python space ever ends up here you'll end up with nagging regarding these.

mwoehlke commented on publish.sh:
```diff
@@ -1,6 +1,6 @@
 #!/bin/sh -e
 
-cd site
+cd _site
```
FYI: okay for now, I guess, but I expect we'll want to follow up by removing this script entirely.

bruxisma replied:
I think we should wait until post-merge before we obliterate the script.

mwoehlke commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+  pull_request:
+    types: [synchronize, edited, opened]
+  push:
+    branches: [main]
```
Er... I may be wrong about how this works, but if this is supposed to be a string match for the branch name, as opposed to `main` being magical:
```suggestion
    branches: [master]
```

(Same in `spellcheck.yml`. And, no, I do not intend to rename the primary branch.)

mwoehlke commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+          --verbose
+          --mode=a+rw
+          --file=cps-docs.tar.gz
+          --directory=${{_site}}
```
Oh. Right, um... stupid question, is this really supposed to be `${{_site}}`, or is it supposed to be `_site`? Or maybe `${{github.workspace}}/_site`? I wonder if you meant to remove the `${{}}` entirely and I, not paying close enough attention, "tricked" you into adding it back... (my bad!)

bruxisma replied:
oh you're totally right. it should be `github.workspace`🤦‍♀️

mwoehlke reviewed (commented):
BTW, please a) rebase, b) squash, c) avoid using emoji in commit messages (at least in summaries!).

The one in the PR summary while it's still WIP is okay.

bruxisma commented:
> BTW, please a) rebase, b) squash, c) avoid using emoji in commit messages (at least in summaries!).
> 
> The one in the PR summary while it's still WIP is okay.

I'll fix the merge conflicts, but I'd recommend you set the following settings for the repository

![image](https://github.com/cps-org/cps/assets/63051/ceb868c8-c853-41c0-b9f6-122907327083)

This would remove the need to worry about merge commits, rebasing, squashing, etc. and would use just the PR description, and not including every small little "wip" or "fix" commit message when merging.

regarding the emoji commit messages, I use [gitmoji](https://gitmoji.dev) so that at a glance someone can have a general idea of what a PR involves, regardless if they are ESL or not. In this case, the PR involves continuous integration, and thus I used the emoji for such. nlohmann of nlohmann json fame is actually the one who inspired me to do this. You can also edit the title of a PR before merging if you'd like 🙂

mwoehlke commented:
> I'd recommend you set the following settings for the repository

I prefer to write my own clean history rather than expecting others to clean up my mess. I also loathe projects that require merge strategies that change the history during the merge. It's akin to maintainers force-pushing to contributor's repositories, and it makes more work for the contributor to clean up afterwards.

...but if *you* want to make your life harder, I won't force you to squash.

As for emoji, I don't necessarily mind them on websites, but git histories are often viewed in terminals where emoji display poorly or even cause issues. And they *aren't* more understandable than text. For example, I'm very much *not* ESL and I took the emoji to mean "work in progress". It seems that isn't what you intended. For some of your other commits, I don't know that I could even begin to guess what meaning is intended.

bruxisma commented:
@bretbrownjr Just got the notification that the workflow was run, but this was before I could handle the other changes and reviews that @mwoehlke had requested. I'll be tweaking this a bit this weekend. 🙂

bruxisma commented:
> > I'd recommend you set the following settings for the repository
> 
> I prefer to write my own clean history rather than expecting others to clean up my mess. I also loathe projects that require merge strategies that change the history during the merge. It's akin to maintainers force-pushing to contributor's repositories, and it makes more work for the contributor to clean up afterwards.
> 
> ...but if _you_ want to make your life harder, I won't force you to squash.

The reason I won't is because I've had cases more often than not where I've caused regressions because of a rebase + squash and then couldn't find the intermediate changes unless I went through every. single. commit. in the pr. It's not fun, and when it comes to devops stuff where debugging is basically "do it live", life is hell for it 😮‍💨 

> As for emoji, I don't necessarily mind them on websites, but git histories are often viewed in terminals where emoji display poorly or even cause issues.

This has not been my experience across all 3 major operating systems, especially windows (cmd.exe doesn't support unicode properly so I don't consider it a terminal of any kind). And, even then if there are issues, all the better to use them to find the bugs lurking in someone's workflow! 😝 

I kid, I kid. That said tools like [`gh`](https://github.com/cli/cli) will strip out emoji for situations like this IIRC. Granted, as I said we can change the PR title which will prevent the PR from going in with an emoji in the first place. I'm afraid to say this *is* a hill I'm willing to die on :v

> And they _aren't_ more understandable than text. For example, I'm very much _not_ ESL and I took the emoji to mean "work in progress". It seems that isn't what you intended. For some of your other commits, I don't know that I could even begin to guess what meaning is intended.

Well I linked the gitmoji website in my comment so you could understand previous commits (minus the lock for lockfile 😓), and have a reference 😄

bretbrownjr commented:
Yeah, apologies. I'm just poking around a bit. Didn't mean to sound the klaxons or anything.

mwoehlke commented:
> I'm afraid to say this *is* a hill I'm willing to die on :v

Likewise. In addition to technical concerns, it's an additional barrier to anyone that hasn't already learned the system, and it's hardly a wide-spread practice. "When in Rome..."

bruxisma commented:
> > I'm afraid to say this _is_ a hill I'm willing to die on :v
> 
> Likewise. In addition to technical concerns, it's an additional barrier to anyone that hasn't already learned the system, and it's hardly a wide-spread practice. "When in Rome..."

"When in Rome..." should be a statement only for those that aren't trying to make any changes in a given location. After all, for hundreds of years humorism was the primary form of , and look how that turned out in the long run 😉

bruxisma commented:
I've edited the PR title to be in line with what you've asked for, merged in latest, and also updated the `Makefile` to rely on the same commands we use, while also supporting some additional features, such as auto-installing the project if no poetry based virtualenv is detected. With the exception of the `publish` target, the Makefile now works on Windows natively as well, and several "builtin" makefile rules and variables are disabled to help when debugging via `make -p` if something goes awry.

@mwoehlke I've also resolved a few conversations with my fixes, and this should be good to go.

mwoehlke commented on Makefile:
```diff
@@ -1,27 +1,48 @@
+.PHONY: all setup cache-clean clean
+
+all: $(if $(venv),,setup)
+	$(POETRY) run sphinx-build ${build.flags} "${SRCDIR}" "${OUTDIR}"
```
BTW, why are some of the variables here using `{}` rather than `()`?
```suggestion
	$(POETRY) run sphinx-build $(build.flags) "$(SRCDIR)" "$(OUTDIR)"
```

bruxisma replied:
I wanted to specify the difference between a command vs a variable. I'm fine with either approach. I just do it out of habit without thinking sometimes (and yes that's why I did `ifeq (${POETRY},)` earlier 😅)

mwoehlke replied:
> I wanted to specify the difference between a command vs a variable. I'm fine with either approach.

Makes sense. However, keeping that consistent seems difficult; IMHO let's just stick with `()` for simplicity's sake. :slightly_smiling_face:

mwoehlke reviewed (changes requested):
This is still not working locally:
```
Running Sphinx v6.2.1

Extension error:
Could not import extension cps (exception: No module named 'cps')
make: *** [Makefile:37: all] Error 2
```

mwoehlke commented:
This isn't working in the GHA environment, either. Same error as local.

bruxisma commented:
> This isn't working in the GHA environment, either. Same error as local.

@mwoehlke Yep, read the error message. Windows case insensitivity strikes again! Just pushed a fix, should work now. 🙂

bruxisma commented:
![image](https://github.com/cps-org/cps/assets/63051/a8a72999-d890-4e19-9955-a41b0e3dd935)
Looks like there's only two warnings and that affects package caching (which can be fixed later. It did build pretty fast without it), BUT the artifact upload step is working. You just need to setup the github pages environment now so the deploy step can work on merge to the default branch 🙂

mwoehlke commented:
> Just pushed a fix, should work now

Joy. Apparently, not only does that break the build, it does so a) without leaving any evidence that's the problem (at least via `make`), and b) once the error has been hit, Poetry is in a state that needs manual repair; just applying the fix doesn't trigger Poetry to sort itself out.

I get the idea of Poetry. I'm becoming less impressed with its reliability.

Also, please see other review comments.

bruxisma commented:
oh that's most likely because I was doing a check against the venv on whether to run the `make setup` target, and that gets created regardless of whether the current package was installed or not. Still, there should be a flag of some kind to fail if the install failed for the current project. At the very least, I could have `make setup` execute a `poetry check` before it runs, as that would have caught the issue. That should also be added to the workflow. I'll add that, then respond to other comments.

mwoehlke commented:
Given that the mere existence of the venv is no guarantee that it is up to date, or even valid, should `all` just depend on `setup` unconditionally? That adds ~1.5s to the build on my machine, but seems worth it for the build to not be broken. (GHA will always need that step, of course, so no difference there.)

While we're at it, maybe the actual page generation should be `html`, with `all` depending on `html` and `html` depending on `setup`... And a target to nuke the Poetry venv would be nice, too...

bruxisma commented:
I didn't want to hoist the cost of an "always install" command on folks without someone else saying they were fine with it. If you're ok with it, then I'm ok with it. 🙂

I can add the venv obliteration as well. I'll add a `purge` target so that you can more or less have a clean setup for the repository, separate from `clean` and `cache-clean`.

mwoehlke commented:
> I didn't want to hoist the cost of an "always install" command on folks without someone else saying they were fine with it

Fair enough. However, given that the cost is ~1.5s versus a potentially broken environment, I don't mind going on record as being okay with it. :slightly_smiling_face: (Anyway, I don't expect many people to be building the HTML locally all that often, so...)

> I'll add a `purge` target...

Thanks!

FWIW, I was thinking `env-clean`/`venv-clean`. :shrug: On its own, "purge" is fine, but w.r.t. the existing `cache-clean` it seems a little less consistent. OTOH, maybe you have a better idea for naming `cache-clean`?

bruxisma commented:
> Oh, sorry, I'm seeing comments in reverse temporal order. ~If there's a way to only require `setup` if `poetry check` is unhappy, that's okay too if that can tell the difference between "a venv exists but is out of date or doesn't actually have stuff 'installed'" and "the venv is ready to go".~ Nope, `poetry check` is happy with a partial install. Looks like unconditionally depending on `setup` / `poetry install` is needed.

We may need to do a check + install as the full `setup` target, or something similar. I'll figure something out.

mwoehlke commented:
> We may need to do a check + install as the full `setup` target

Note that, as far as I can tell, `poetry check` reports "All set!" even if the local project isn't installed:

```
$ poetry install
...
[Errno 2] No such file or directory: '/home/src/matthew/work/cps/README.md'
$ poetry check
All set!
```

(I intentionally reintroduced the readme issue so the local project install wouldn't happen.)

That being the case, unless there's a more "complete" test, I think HTML generation should just depend on `poetry install` unconditionally.

bruxisma commented:
Sorry for not responding sooner. Busy day :v

I've set the `setup` target to be a dependency of `all`, add a `purge` target that cleans the venv and the build directory, changed the target names to `clean.venv` and `clean.cache`, but kept `clean` for the build itself. I've also set it up so that we always check the pyproject.toml config in CI and during the build as well just in case. When you're in a broken state, there's no way for us to run install then check and have that work. But we could do check + install and that would prevent a failed install. You were in a broken state because of out of order execution. 

I also added your code suggestion. I'll get to the rest of this in the AM. 🫡

bruxisma commented on Makefile:
```diff
@@ -1,27 +1,48 @@
-ifeq ($(shell which $(SPHINXBUILD) >/dev/null 2>&1; echo $$?), 1)
-$(error The '$(SPHINXBUILD)' command was not found. Make sure you have Sphinx installed, then set the SPHINXBUILD environment variable to point to the full path of the '$(SPHINXBUILD)' executable. Alternatively you can add the directory with the executable to your PATH. If you don't have Sphinx installed, grab it from http://sphinx-doc.org/)
+ifeq (${POETRY},)
+$(error Could not find the `poetry` command. \
+	Please make sure you have installed poetry, and that it is on your system's PATH environment variable. \
+	If you don't have poetry installed, please visit https://python-poetry.org for instructions on its installation.)
```
That works for me!

bruxisma commented on .github/workflows/docs.yml:
```diff
@@ -0,0 +1,57 @@
+      - name: Install Poetry
+        run: pipx install poetry
+      - name: Setup Python
+        uses: actions/setup-python@v5
+        with:
+          cache: poetry
```
We can also set it to `3.x` and it would just use the latest. I'm not TOO worried about a Python 4.0 happening anytime soon *hears the gnashing of teeth of python developers in the year 2008*

mwoehlke replied:
I like 3.x! :+1: (I didn't know that was an option...)

mwoehlke commented:
@bruxisma, unless I hear an objection Very Soon, I am going to do two things:

1. Push some minor fixes to fully resolve the issues I see. (Mostly, using `make` in the GHA recipe and some very minor  issues in the `Makefile`.)
2. Replace the description in this Issue with the intended commit message, as it's still not obvious if GitHub is going to let me edit that beforehand.

For posterity, here is the original description:
```md
This PR closes #40 and brings in a few changes to the general build workflow:

1. [`poetry`][1] is now used to setup a virtual environment separate from your global environment to allow having a uniform installation of tools.
2. `cps` now exists as a sphinx plugin + module and is "installable". This also applies for the forked/modified `autosectionlabel`. Details on how to add additional custom extensions can be found in the `pyproject.toml` file (`tool.poetry.packages` and the `tool.poetry.plugins."sphinx-builders"` sections)
3. We do NOT use the github [upload-pages-artifact][2] action, as we can package + fixup permissions in one move.
4. We do NOT use the Makefile during the CI step.
5. We cache sphinx dependencies now so when upgrading/installing things should go much faster (especially on PRs)

GitHub Pages *do not yet support* previews of PRs for the general public. However, GitHub Pages alternatives such as Cloudflare Pages *do*. This is something I'd recommend considering if having per-commit + per-branch previews is desired.

> [!TIP]
> I'd recommend *possibly* putting some approvals for the actual final deployment to GitHub Pages in place

Lastly, I took an extra step and added a spellchecker for the github actions steps. Unfortunately, they do not currently support having generalized tags for a specific version. Keeping up to date with it without automation would be difficult. For this reason, we're currently using `@master`. However, I'd recommend considering installing [renovatebot][3] to help keep our github actions workflows up to date and also *pinned* to specific git commit hashes. 

You can see how this operates in practice by looking at [my personal `workflow_call` capable workflows][4] and how they've managed to keep actions pinned and up to date for *quite some time*. The usage of renovatebot requires the installation at the organization level, and so I'm unable to actually turn it on for CPS itself 🙂

> [!NOTE]
> I understand this is a lot of suggestions + changes, and also this initial PR might shit the bed right out of the gate, but we can get that resolved ASAP.

[1]: https://python-poetry.org
[2]: https://github.com/actions/upload-pages-artifact/
[3]: https://docs.renovatebot.com/
[4]: https://github.com/bruxisma/actions/pulls?q=is%3Apr+is%3Aclosed
```

bruxisma commented:
> @bruxisma, unless I hear an objection Very Soon, I am going to do two things:
> 
>     1. Push some minor fixes to fully resolve the issues I see. (Mostly, using `make` in the GHA recipe and some very minor  issues in the `Makefile`.)
> 
>     2. Replace the description in this Issue with the intended commit message, as it's still not obvious if GitHub is going to let me edit that beforehand.

Alas, I had a late morning and didn't get to a PC until now. I'll get the last things fixed up in about 20-30 minutes.

mwoehlke commented:
> I'll get the last things fixed up in about 20-30 minutes.

Ack, you managed to post that *seconds* after I pushed fixes. I'm okay with the current state; is there anything else *you* think still needs attention?

(Besides dropping the `@` from the `make setup` commands — I don't think we need that — the `Makefile` changes are hopefully straight-forward. Mostly it was rearranging things, fixing the `.PHONY` targets, removing the unused `srcdir` and adding `setup.flags`.)

mwoehlke commented:
Also, please let me know if you think anything is missing in the description. I took what I'd written for #49, but there are additional refinements here.

bruxisma commented:
I updated the description slightly to add more context and use a bit more markdown while not destroying the footnotes you'd used, but also not using the `[word](link)` syntax.

One last thing I think is worth doing: adding an `archive` or `prepare-archive` task so we can share steps between CI and local for testing. Thoughts? Wouldn't take more than a few minutes. `tar` has been available on Windows since about 2018 under system32, so it shows up on the system path. (It's also `bsdtar`, the same `tar` used by macOS)

mwoehlke commented:
> One last thing I think is worth doing: adding an `archive` or `prepare-archive` task so we can share steps between CI and local for testing. Thoughts?

I was actually just thinking the exact same thing. I was planning to circle back so as to not be pushing further changes to your branch, but since you had the same thought, let's do it. I vote for `archive`, though, and preferably put it right before `publish`; thanks! :slightly_smiling_face:

bruxisma commented:
Will do! This way since I'll be pushing the changes, we won't have to wait for bret to approve before merge 😝

mwoehlke commented:
Well, I _don't_ have to wait on Bret, though I pinged him to see if he wants to do a review anyway. If I don't hear back, I'll wait until tomorrow to press the button.

Thanks for the cross-platform note! I changed a couple words to keep the tone more consistent, but that's indeed a notable mention I'd overlooked. Since this is intended to become the commit message verbatim, I prefer to avoid markdown that doesn't significantly benefit a reader that only sees plaintext, but I can live with that mechanism for footnotes. :slightly_smiling_face: I also re-wrapped for the same reason (i.e. because I don't trust GitHub to do it right), and I preemptively added a few words about archive generation.

bretbrownjr commented:
I won't be able to review until tomorrow in the Western hemisphere, but I'm not worried about that. If you two are satisfied, I am too.

mwoehlke commented:
> If you two are satisfied, I am too.

Okay, thanks @bretbrownjr!

mwoehlke commented on Makefile:
```diff
@@ -1,27 +1,69 @@
+venv := $(notdir $(realpath $(shell $(POETRY) env info --path 2> $(--null))))
+
+archive.flags += --create --gzip --verbose
+archive.flags += --file=$(if $(ARCHIVE),$(ARCHIVE),cps-docs).tar.gz
```
If the user specifies `ARCHIVE`, wouldn't that be expected to include the extension?

Otherwise LGTM, thanks!

bruxisma replied:
possibly. I could do the extra work of suffix detection/adding via `make`'s builtins for it if you think that's worth the time. Downside is technically someone could do `make archive ARCHIVE=docs.tar.bz2` and it wouldn't be a `bz2` 😅

mwoehlke replied:
IM(NS)HO there's no need to get that fancy. We generate a gzip'd tar; if you supply an `ARCHIVE` with an inappropriate extensions, that's on you. :slightly_smiling_face: I think it's less surprising though for `ARCHIVE` to be the full name and not just the name-without-extension. (Also, `ARCHIVE=foo.tgz` is fine.)

mwoehlke replied:
To make my position clearer :wink::
```suggestion
archive.flags += --file=$(if $(ARCHIVE),$(ARCHIVE),cps-docs.tar.gz)
```

bruxisma replied:
Done and done then 🙂

mwoehlke commented:
...and I belatedly realize that `$(if $(foo),$(foo),stuff)` can be replaced with `FOO ?= stuff`, so I did that. :slightly_smiling_face: 

Last call for pre-merge objections! :tada:

bruxisma commented:
One last commit, since you simplified the `if` 😝

mwoehlke commented:
Heh, I didn't even notice that one. :slightly_smiling_face: :+1:

mwoehlke commented:
I'll take that as a lack of objections, then!

mwoehlke commented:
Okay, for future reference, I *can* edit the commit message after pressing the button. (Not total wasted effort, though, since having done so preemptively, I didn't *need* to edit it...)

Thanks again!

bruxisma commented:
welp, sad news: deploy step failed, and the error message wasn't too helpful. I assume maybe the environment or similar isn't setup? 😬

mwoehlke commented:
@bruxisma, yes, I was going to ask if you have any thoughts. Default(?) workflow permissions are read-only but the option to change that is disabled (and probably not what we want, anyway). The `github-pages` environment exists and seems to be correctly configured. Pages source is already GHA (which a) surprised me, and b) makes me wonder how the old pages, which AFAIK are using the classic "branch" model, are still present).

I don't recall making changes during the course of this PR because it's not clear exactly what needs to change? Maybe you have some ideas?

bruxisma commented:
I'll look into it shortly. I've got some fires to put out at `$JOB` I'm afraid :(

mwoehlke commented:
@bruxisma, solved! And I think the solution may not have been previously known. (Also, that was fun; the only way to test was to push a hope-this-works commit directly to master. I was going to force-push that away real fast if it *hadn't* worked!)

bruxisma commented:
*oof*. Glad it worked at least. One of the things I like about cloudflare pages, vercel, et. al. is they allow previews in case deployment to master/main fails, whereas that's not available to github pages yet and I don't know what the roadmap for it is.

mwoehlke commented:
Right, though the problem being with deploy-pages, I'm not sure if that would have helped in this instance. :slightly_smiling_face: In any case, I left an annoyed comment that upload-pages-artifact really ought to fix their misleading documentation.

I'm about to push a follow-up PR to fix some minor things I noticed.

mwoehlke commented:
#54 if you're interested...
